ckanext/subscribe/utils.py

```diff
@@ -0,0 +1,227 @@
+# encoding: utf-8
+from jinja2 import Template
+
+import ckan.plugins as p
+from ckan import model
+
+config = p.toolkit.config
+
+
+def get_footer_contents(email_vars, subscription=None):
+    html_lines = []
+    if subscription:
+        html_lines.append(
+            'To stop receiving emails of this type: '
+            '<a href="{unsubscribe_link}">'
+            'unsubscribe from {object_type} "{object_title}"</a>'
+        )
+    else:
+        html_lines.append(
+            'To stop receiving all subscription emails from {site_title}: '
+            '<a href="{unsubscribe_all_link}">'
+            'unsubscribe all</a>'
+        )
+    html_lines.append('<a href="{manage_link}">Manage settings</a>')
+    html_footer = '\n'.join(
+        '<p style="font-size:10px;line-height:200%;text-align:center;'
+        'color:#9EA3A8=;padding-top:0px">{line}</p>'.format(line=line)
+        for line in html_lines).format(**email_vars)
+
+    plain_text_footer = ''
+    if subscription:
+        plain_text_footer += '''
+You can unsubscribe from notifications emails for {object_type}: "{object_title}" by going to {unsubscribe_link}.
+'''.format(**email_vars)
+    else:
+        plain_text_footer += (
+            'To stop receiving all subscription emails from {site_title}: '
+            '<a href="{unsubscribe_all_link}">'
+            'unsubscribe all</a>'
+        ).format(**email_vars)
+    plain_text_footer += '''
+Manage your settings at {manage_link}.
+'''.format(**email_vars)
+    return plain_text_footer, html_footer
+
+
+def get_email_vars(code, subscription=None, email=None):
+    '''
+    Get the variables to substitute into email templates.
+
+    If you have a subscription object, not just an email address, then you get
+    the object_* variables which allow you to link to the object subscribed to.
+
+    :param code: the email auth code (required)
+    :param subscription: subscription object (optional)
+    :param email: email address of the user (supply a value if subscription is
+        not available)
+    '''
+    assert code
+    assert subscription or email
+    unsubscribe_all_link = p.toolkit.url_for(
+        controller='ckanext.subscribe.controller:SubscribeController',
+        action='unsubscribe_all',
+        code=code,
+        qualified=True)
+    manage_link = p.toolkit.url_for(
+        controller='ckanext.subscribe.controller:SubscribeController',
+        action='manage',
+        code=code,
+        qualified=True)
+    extra_vars = dict(
+        site_title=config.get('ckan.site_title'),
+        site_url=config.get('ckan.site_url'),
+        email=email or subscription.email,
+        unsubscribe_all_link=unsubscribe_all_link,
+        manage_link=manage_link,
```

Comment:
```suggestion
        manage_link=manage_link
```

Reply:
The trailing comma doesn't make a difference here, and I prefer to leave them in for multi-line statements like this, because it minimises the number of line changes in git if we make an edit later.

Reply:
fine to me